outlines/generate/generator.py

```diff
@@ -0,0 +1,237 @@
+import math
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Callable, List, Optional, Union
+
+import torch
+
+if TYPE_CHECKING:
+    from outlines.generate.samplers import Sampler
+    from outlines.index.index import Index
+
+"""
+--------------------
+Functions vs classes
+--------------------
+
+def text_generator(model, tokenizer, sampler, fsm):
+    generate_token = token_generator(model, sampler)
+    generator = generator(generate_token, fsm, tokenizer)
+    return generator
+
+generator = text_generator(model, tokenizer, sampler, fsm)
+generator("prompt")
+
+def call_text_generator(prompt):
+    _*, sequence = generator(prompt)
+    return sequence
+
+--> We need to chain the generator with a generator that yields a `Sequence` state.
+--> Then use this generator at the end when calling the calling function directly
+
+def json(model, tokenizer, sampler, fsm):
+    generate_token = token_generator(model, sampler)
+    generator = sequence_generator(generate_token, fsm, tokenizer)
+    return generator
+
+"""
+
+
+@dataclass(frozen=True)
+class GenerationState:
+    token_ids: torch.Tensor
+    attention_masks: torch.Tensor
+    kv_cache: Optional[torch.Tensor] = None
+    rng: Optional[torch.Generator] = None
+
+
+class SequenceGenerator:
+    def __init__(self, fsm, model, sampler):
+        self.generate_token = token_generator(model, sampler)
+        self.fsm = fsm
+        self.model = model
+        self.device = self.model.device
+
+    def init_generation_state(
+        self, prompt: Union[str, List[str]], rng: Optional[torch.Generator] = None
+    ):
+        """Initialize the generation state.
+
+        This method is responsible for encoding the prompt, moving token ids
+        to the device and initializing the random number generator.
+
+        Parameters
+        ----------
+        prompt
+            The prompt on which the generation is conditioned.
+        rng
+            The state of the random number generator.
+
+        Returns
+        -------
+        A `GenerationState` object.
+
+        """
+        token_ids, attention_masks = self.model.tokenizer.encode(prompt)
```

Comment:
Also `self.model` necessary because of this function. It needs to be factored out.